crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl

```diff
@@ -0,0 +1,195 @@
+    }
+
+    workgroupBarrier();
+
+    // Load 1 triangle's worth of vertex data per thread
+    let triangle_id = local_invocation_index;
+    if triangle_id >= meshlet.triangle_count { return; }
+    let index_ids = meshlet.start_index_id + (triangle_id * 3u) + vec3(0u, 1u, 2u);
+    let vertex_ids = vec3(get_meshlet_index(index_ids[0]), get_meshlet_index(index_ids[1]), get_meshlet_index(index_ids[2]));
+    let vertex_0 = viewport_vertices[vertex_ids[2]];
+    let vertex_1 = viewport_vertices[vertex_ids[1]];
+    let vertex_2 = viewport_vertices[vertex_ids[0]];
+    let packed_ids = (cluster_id << 6u) | triangle_id;
+
+    // Compute triangle bounding box
+    let min_x = u32(min3(vertex_0.x, vertex_1.x, vertex_2.x));
+    let min_y = u32(min3(vertex_0.y, vertex_1.y, vertex_2.y));
+    var max_x = u32(ceil(max3(vertex_0.x, vertex_1.x, vertex_2.x)));
+    var max_y = u32(ceil(max3(vertex_0.y, vertex_1.y, vertex_2.y)));
+    max_x = min(max_x, u32(view.viewport.z) - 1u);
+    max_y = min(max_y, u32(view.viewport.w) - 1u);
+    if any(vec2(min_x, min_y) > vec2(max_x, max_y)) { return; }
+
+    // Setup triangle gradients
+    let w_x = vec3(vertex_1.y - vertex_2.y, vertex_2.y - vertex_0.y, vertex_0.y - vertex_1.y);
+    let w_y = vec3(vertex_2.x - vertex_1.x, vertex_0.x - vertex_2.x, vertex_1.x - vertex_0.x);
+    let triangle_double_area = edge_function(vertex_0.xy, vertex_1.xy, vertex_2.xy); // TODO: Reuse earlier calculations and take advantage of summing to 1
+    let vertices_z = vec3(vertex_0.z, vertex_1.z, vertex_2.z) / triangle_double_area;
+    let z_x = dot(vertices_z, w_x);
+    let z_y = dot(vertices_z, w_y);
+
+    // Setup initial triangle equations
+    let starting_pixel = vec2(f32(min_x), f32(min_y)) + 0.5;
+    var w_row = vec3(
+        // TODO: Reuse earlier calculations and take advantage of summing to 1
+        edge_function(vertex_1.xy, vertex_2.xy, starting_pixel),
+        edge_function(vertex_2.xy, vertex_0.xy, starting_pixel),
+        edge_function(vertex_0.xy, vertex_1.xy, starting_pixel),
+    );
+    var z_row = dot(vertices_z, w_row);
+    let view_width = u32(view.viewport.z);
+    var frag_coord_1d_row = min_y * view_width;
+
+    // Rasterize triangle
+    if subgroupAny(max_x - min_x > 4u) {
+        // Scanline setup
+        let edge_012 = -w_x;
+        let open_edge = edge_012 < vec3(0.0);
+        let inverse_edge_012 = select(1.0 / edge_012, vec3(1e8), edge_012 == vec3(0.0));
+        let max_x_diff = vec3<f32>(max_x - min_x);
+        for (var y = min_y; y <= max_y; y++) {
+            // Calculate start and end X interval for pixels in this row within the triangle
+            let cross_x = w_row * inverse_edge_012;
+            let min_x2 = select(vec3(0.0), cross_x, open_edge);
+            let max_x2 = select(cross_x, max_x_diff, open_edge);
+            var x0 = u32(ceil(max3(min_x2[0], min_x2[1], min_x2[2])));
+            let x1 = u32(min3(max_x2[0], max_x2[1], max_x2[2])) + min_x;
```

Comment:
i'd move this `+ min_x` down to be next to the other one just so its clearer that both x0 and x1 are both shifted

Reply:
Done